File: backend/app/schemas/invoice_match.py
from __future__ import annotations
from datetime import date, datetime
from decimal import Decimal
from typing import List, Optional, Any
from uuid import UUID

from pydantic import BaseModel, ConfigDict


# ── Tolerance Rule ────────────────────────────────────────────────────────────

class ToleranceRuleCreate(BaseModel):
    rule_name: str
    supplier_id: Optional[UUID] = None
    item_category: Optional[str] = None
    quantity_tolerance_pct: Decimal = Decimal("1")
    quantity_tolerance_abs: Optional[Decimal] = None
    price_tolerance_pct: Decimal = Decimal("2")
    price_tolerance_abs: Optional[Decimal] = None
    tax_tolerance_pct: Decimal = Decimal("0")
    auto_approve_within_tolerance: bool = True
    block_payment_beyond_tolerance: bool = True
    approval_role_required: Optional[str] = None
    is_active: bool = True
    priority: int = 10
    notes: Optional[str] = None


class ToleranceRuleUpdate(BaseModel):
    rule_name: Optional[str] = None
    quantity_tolerance_pct: Optional[Decimal] = None
    price_tolerance_pct: Optional[Decimal] = None
    tax_tolerance_pct: Optional[Decimal] = None
    auto_approve_within_tolerance: Optional[bool] = None
    block_payment_beyond_tolerance: Optional[bool] = None
    is_active: Optional[bool] = None
    priority: Optional[int] = None
    notes: Optional[str] = None


class ToleranceRuleOut(ToleranceRuleCreate):
    model_config = ConfigDict(from_attributes=True)
    id: UUID
    created_at: datetime
    updated_at: datetime


# ── GRN Link ──────────────────────────────────────────────────────────────────

class MatchGRNLinkOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    id: UUID
    grn_id: Optional[UUID] = None
    grn_line_id: Optional[UUID] = None
    grn_no: Optional[str] = None
    received_qty: Optional[Decimal] = None
    accepted_qty: Optional[Decimal] = None
    rejected_qty: Optional[Decimal] = None
    received_date: Optional[date] = None


# ── Match Line ────────────────────────────────────────────────────────────────

class MatchLineOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    id: UUID
    header_id: UUID
    invoice_line_id: Optional[UUID] = None
    po_line_id: Optional[UUID] = None
    material_id: Optional[UUID] = None
    material_name: Optional[str] = None
    description: Optional[str] = None
    ordered_qty: Optional[Decimal] = None
    ordered_unit_price: Optional[Decimal] = None
    ordered_tax_rate: Optional[Decimal] = None
    total_received_qty: Optional[Decimal] = None
    total_accepted_qty: Optional[Decimal] = None
    prev_invoiced_qty: Decimal
    billable_qty: Optional[Decimal] = None
    invoiced_qty: Optional[Decimal] = None
    invoiced_unit_price: Optional[Decimal] = None
    invoiced_tax_rate: Optional[Decimal] = None
    invoiced_line_total: Optional[Decimal] = None
    qty_variance: Optional[Decimal] = None
    price_variance: Optional[Decimal] = None
    tax_variance: Optional[Decimal] = None
    qty_variance_pct: Optional[Decimal] = None
    price_variance_pct: Optional[Decimal] = None
    match_status: str
    tolerance_rule_id: Optional[UUID] = None
    action_required: bool
    notes: Optional[str] = None
    grn_links: List[MatchGRNLinkOut] = []


# ── Review ────────────────────────────────────────────────────────────────────

class ReviewCreate(BaseModel):
    action: str
    reviewer_name: str
    reviewer_role: Optional[str] = None
    justification: Optional[str] = None
    line_ids: Optional[List[str]] = None


class ReviewOut(ReviewCreate):
    model_config = ConfigDict(from_attributes=True)
    id: UUID
    header_id: UUID
    previous_status: Optional[str] = None
    new_status: Optional[str] = None
    created_at: datetime


# ── Header ────────────────────────────────────────────────────────────────────

class MatchHeaderSummary(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    id: UUID
    match_no: str
    invoice_id: UUID
    po_id: Optional[UUID] = None
    supplier_id: Optional[UUID] = None
    invoice_no: Optional[str] = None
    invoice_date: Optional[date] = None
    currency: str
    invoice_total: Decimal
    overall_status: str
    duplicate_flag: bool
    review_required: bool
    is_payable: bool
    payment_hold_reason: Optional[str] = None
    run_at: Optional[datetime] = None
    created_at: datetime
    updated_at: datetime


class MatchHeaderOut(MatchHeaderSummary):
    reviewed_by: Optional[str] = None
    reviewed_at: Optional[datetime] = None
    approved_by: Optional[str] = None
    approved_at: Optional[datetime] = None
    notes: Optional[str] = None
    match_lines: List[MatchLineOut] = []
    reviews: List[ReviewOut] = []


# ── Duplicate Log ─────────────────────────────────────────────────────────────

class DuplicateLogOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    id: UUID
    invoice_id: UUID
    matched_invoice_id: Optional[UUID] = None
    invoice_no: Optional[str] = None
    matched_invoice_no: Optional[str] = None
    duplicate_risk: str
    match_reasons: Optional[Any] = None
    similarity_score: Optional[Decimal] = None
    is_resolved: bool
    resolved_by: Optional[str] = None
    resolved_at: Optional[datetime] = None
    resolution_notes: Optional[str] = None
    created_at: datetime


# ── AI Rec ────────────────────────────────────────────────────────────────────

class IMAIRecOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    id: UUID
    header_id: Optional[UUID] = None
    agent_type: str
    status: str
    title: str
    explanation: Optional[str] = None
    impact_summary: Optional[str] = None
    suggested_action: Optional[str] = None
    priority: str
    confidence_score: Optional[Decimal] = None
    actioned_at: Optional[datetime] = None
    created_at: datetime


# ── Dashboard ─────────────────────────────────────────────────────────────────

class IMDashboard(BaseModel):
    total_matches: int
    fully_matched: int
    within_tolerance: int
    mismatched: int
    blocked: int
    pending_review: int
    duplicate_suspected: int
    payable_count: int
    on_hold_count: int
    pending_ai_recs: int
    mismatch_by_type: List[dict]
    recent_matches: List[MatchHeaderSummary]


# ── Run Request ───────────────────────────────────────────────────────────────

class RunMatchRequest(BaseModel):
    invoice_id: UUID
    force_rerun: bool = False

File: frontend/src/lib/invoice_match.ts
import { apiClient } from "@/lib/api";

// ── Enums / Types ─────────────────────────────────────────────────────────────

export type MatchLineStatus =
  | "FULLY_MATCHED" | "MATCHED_WITHIN_TOLERANCE"
  | "QUANTITY_MISMATCH" | "PRICE_MISMATCH" | "TAX_MISMATCH"
  | "MISSING_PO" | "MISSING_GRN" | "PARTIAL_MATCH" | "OVERBILLED"
  | "BLOCKED" | "APPROVED_EXCEPTION" | "REJECTED";

export type MatchOverallStatus =
  | "FULLY_MATCHED" | "MATCHED_WITHIN_TOLERANCE" | "PARTIAL_MATCH"
  | "QUANTITY_MISMATCH" | "PRICE_MISMATCH" | "MISSING_PO" | "MISSING_GRN"
  | "DUPLICATE_SUSPECTED" | "BLOCKED" | "APPROVED_EXCEPTION"
  | "REJECTED" | "PENDING_REVIEW";

export type IMAIAgentType = "ANOMALY_DETECTOR" | "EXCEPTION_PRIORITIZER" | "TOLERANCE_TUNER";
export type IMAIRecStatus = "PENDING" | "ACCEPTED" | "REJECTED";
export type DuplicateRisk = "EXACT" | "SUSPECTED" | "LOW";

// ── Interfaces ────────────────────────────────────────────────────────────────

export interface ToleranceRule {
  id: string;
  rule_name: string;
  supplier_id: string | null;
  item_category: string | null;
  quantity_tolerance_pct: number;
  quantity_tolerance_abs: number | null;
  price_tolerance_pct: number;
  price_tolerance_abs: number | null;
  tax_tolerance_pct: number;
  auto_approve_within_tolerance: boolean;
  block_payment_beyond_tolerance: boolean;
  approval_role_required: string | null;
  is_active: boolean;
  priority: number;
  notes: string | null;
  created_at: string;
  updated_at: string;
}

export interface MatchGRNLink {
  id: string;
  grn_id: string | null;
  grn_line_id: string | null;
  grn_no: string | null;
  received_qty: number | null;
  accepted_qty: number | null;
  rejected_qty: number | null;
  received_date: string | null;
}

export interface MatchLine {
  id: string;
  header_id: string;
  invoice_line_id: string | null;
  po_line_id: string | null;
  material_id: string | null;
  material_name: string | null;
  description: string | null;
  ordered_qty: number | null;
  ordered_unit_price: number | null;
  ordered_tax_rate: number | null;
  total_received_qty: number | null;
  total_accepted_qty: number | null;
  prev_invoiced_qty: number;
  billable_qty: number | null;
  invoiced_qty: number | null;
  invoiced_unit_price: number | null;
  invoiced_tax_rate: number | null;
  invoiced_line_total: number | null;
  qty_variance: number | null;
  price_variance: number | null;
  tax_variance: number | null;
  qty_variance_pct: number | null;
  price_variance_pct: number | null;
  match_status: MatchLineStatus;
  tolerance_rule_id: string | null;
  action_required: boolean;
  notes: string | null;
  grn_links: MatchGRNLink[];
}

export interface MatchReview {
  id: string;
  header_id: string;
  action: string;
  reviewer_name: string;
  reviewer_role: string | null;
  previous_status: string | null;
  new_status: string | null;
  justification: string | null;
  created_at: string;
}

export interface MatchHeaderSummary {
  id: string;
  match_no: string;
  invoice_id: string;
  po_id: string | null;
  supplier_id: string | null;
  invoice_no: string | null;
  invoice_date: string | null;
  currency: string;
  invoice_total: number;
  overall_status: MatchOverallStatus;
  duplicate_flag: boolean;
  review_required: boolean;
  is_payable: boolean;
  payment_hold_reason: string | null;
  run_at: string | null;
  created_at: string;
  updated_at: string;
}

export interface MatchHeaderOut extends MatchHeaderSummary {
  reviewed_by: string | null;
  reviewed_at: string | null;
  approved_by: string | null;
  approved_at: string | null;
  notes: string | null;
  match_lines: MatchLine[];
  reviews: MatchReview[];
}

export interface DuplicateLog {
  id: string;
  invoice_id: string;
  matched_invoice_id: string | null;
  invoice_no: string | null;
  matched_invoice_no: string | null;
  duplicate_risk: DuplicateRisk;
  match_reasons: string[] | null;
  similarity_score: number | null;
  is_resolved: boolean;
  resolved_by: string | null;
  resolved_at: string | null;
  resolution_notes: string | null;
  created_at: string;
}

export interface IMAIRec {
  id: string;
  header_id: string | null;
  agent_type: IMAIAgentType;
  status: IMAIRecStatus;
  title: string;
  explanation: string | null;
  impact_summary: string | null;
  suggested_action: string | null;
  priority: string;
  confidence_score: number | null;
  actioned_at: string | null;
  created_at: string;
}

export interface IMDashboard {
  total_matches: number;
  fully_matched: number;
  within_tolerance: number;
  mismatched: number;
  blocked: number;
  pending_review: number;
  duplicate_suspected: number;
  payable_count: number;
  on_hold_count: number;
  pending_ai_recs: number;
  mismatch_by_type: { type: string; count: number }[];
  recent_matches: MatchHeaderSummary[];
}

// ── Color / Label maps ────────────────────────────────────────────────────────

export const MATCH_STATUS_COLOR: Record<MatchOverallStatus, string> = {
  FULLY_MATCHED:            "bg-green-100 text-green-700",
  MATCHED_WITHIN_TOLERANCE: "bg-teal-100 text-teal-700",
  PARTIAL_MATCH:            "bg-yellow-100 text-yellow-700",
  QUANTITY_MISMATCH:        "bg-orange-100 text-orange-700",
  PRICE_MISMATCH:           "bg-red-100 text-red-700",
  MISSING_PO:               "bg-red-200 text-red-800",
  MISSING_GRN:              "bg-red-200 text-red-800",
  DUPLICATE_SUSPECTED:      "bg-purple-100 text-purple-700",
  BLOCKED:                  "bg-red-600 text-white",
  APPROVED_EXCEPTION:       "bg-blue-100 text-blue-700",
  REJECTED:                 "bg-gray-300 text-gray-700",
  PENDING_REVIEW:           "bg-yellow-200 text-yellow-800",
};

export const LINE_STATUS_COLOR: Record<MatchLineStatus, string> = {
  FULLY_MATCHED:            "bg-green-100 text-green-700",
  MATCHED_WITHIN_TOLERANCE: "bg-teal-100 text-teal-700",
  QUANTITY_MISMATCH:        "bg-orange-100 text-orange-700",
  PRICE_MISMATCH:           "bg-red-100 text-red-700",
  TAX_MISMATCH:             "bg-amber-100 text-amber-700",
  MISSING_PO:               "bg-red-200 text-red-800",
  MISSING_GRN:              "bg-rose-100 text-rose-700",
  PARTIAL_MATCH:            "bg-yellow-100 text-yellow-700",
  OVERBILLED:               "bg-red-600 text-white",
  BLOCKED:                  "bg-red-600 text-white",
  APPROVED_EXCEPTION:       "bg-blue-100 text-blue-700",
  REJECTED:                 "bg-gray-300 text-gray-700",
};

export const AI_PRIORITY_COLOR: Record<string, string> = {
  LOW:      "bg-gray-100 text-gray-600",
  MEDIUM:   "bg-yellow-100 text-yellow-700",
  HIGH:     "bg-orange-100 text-orange-700",
  CRITICAL: "bg-red-100 text-red-700",
};

export const AI_AGENT_LABEL: Record<IMAIAgentType, string> = {
  ANOMALY_DETECTOR:      "Anomaly Detector",
  EXCEPTION_PRIORITIZER: "Exception Prioritizer",
  TOLERANCE_TUNER:       "Tolerance Tuner",
};

export function varColor(pct: number | null): string {
  if (pct === null) return "text-gray-400";
  const abs = Math.abs(pct);
  if (abs === 0) return "text-green-600";
  if (abs <= 2) return "text-yellow-600";
  if (abs <= 5) return "text-orange-600";
  return "text-red-600";
}

export function fmtVar(v: number | null, suffix = ""): string {
  if (v === null || v === undefined) return "—";
  const sign = v > 0 ? "+" : "";
  return `${sign}${v.toFixed(2)}${suffix}`;
}

// ── API ───────────────────────────────────────────────────────────────────────

const BASE = "/api/v1/invoice-match";

export const imApi = {
  getDashboard: () => apiClient.get<IMDashboard>(`${BASE}/dashboard`),

  // Matching
  runMatch: (invoice_id: string, force_rerun = false) =>
    apiClient.post<MatchHeaderOut>(`${BASE}/run`, { invoice_id, force_rerun }),

  // List / Get
  list: (params?: { status?: string; review_required?: boolean; duplicate_flag?: boolean }) => {
    const qs = new URLSearchParams(
      Object.entries(params || {})
        .filter(([, v]) => v !== undefined && v !== null)
        .map(([k, v]) => [k, String(v)])
    ).toString();
    return apiClient.get<MatchHeaderSummary[]>(`${BASE}${qs ? `?${qs}` : ""}`);
  },
  get: (id: string) => apiClient.get<MatchHeaderOut>(`${BASE}/${id}`),
  getBlockedInvoices: () => apiClient.get<MatchHeaderSummary[]>(`${BASE}/blocked-invoices`),
  getDuplicateSuspicions: () => apiClient.get<DuplicateLog[]>(`${BASE}/duplicate-suspicions`),

  // Review
  review: (id: string, body: Record<string, unknown>) =>
    apiClient.post<MatchHeaderOut>(`${BASE}/${id}/review`, body),
  approveException: (id: string, reviewer_name: string, justification: string) =>
    apiClient.post<MatchHeaderOut>(`${BASE}/${id}/approve-exception`, { reviewer_name, justification }),
  reject: (id: string, reviewer_name: string, justification: string) =>
    apiClient.post<MatchHeaderOut>(`${BASE}/${id}/reject`, { reviewer_name, justification }),

  // Duplicates
  resolveDuplicate: (dupId: string, resolved_by: string, resolution_notes: string) =>
    apiClient.post<DuplicateLog>(`${BASE}/duplicate-suspicions/${dupId}/resolve`, { resolved_by, resolution_notes }),

  // Tolerance rules
  listToleranceRules: (active_only = false) =>
    apiClient.get<ToleranceRule[]>(`${BASE}/tolerance-rules${active_only ? "?active_only=true" : ""}`),
  createToleranceRule: (body: Record<string, unknown>) =>
    apiClient.post<ToleranceRule>(`${BASE}/tolerance-rules`, body),
  updateToleranceRule: (id: string, body: Record<string, unknown>) =>
    apiClient.patch<ToleranceRule>(`${BASE}/tolerance-rules/${id}`, body),

  // Reports
  reportMismatchSummary: () =>
    apiClient.get<Record<string, unknown>[]>(`${BASE}/reports/mismatch-summary`),
  reportSupplierAccuracy: () =>
    apiClient.get<Record<string, unknown>[]>(`${BASE}/reports/supplier-accuracy`),

  // AI
  runAI: () => apiClient.post<{ recommendations_created: number }>(`${BASE}/ai/run`, {}),
  runAIForMatch: (id: string) => apiClient.post<{ recommendations_created: number }>(`${BASE}/${id}/ai/run`, {}),
  listAIRecs: (params?: { status?: string }) => {
    const qs = params?.status ? `?status=${params.status}` : "";
    return apiClient.get<IMAIRec[]>(`${BASE}/ai/recs${qs}`);
  },
  actionAIRec: (recId: string, status: "ACCEPTED" | "REJECTED") =>
    apiClient.post<IMAIRec>(`${BASE}/ai/recs/${recId}/action`, { status }),
};
